### DISCUSSION TITLE
[JS] 익명 함수

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#86] [프론트 기술 질문 공간]
JavaScript에서 익명 함수란 무엇이고, 익명 함수의 일반적인 사용 사례는 무엇인가요?

**Comment by michoball** [comment]
일반함수처럼 매개변수를 받고 새로운 값은 표출하고 할 수 있지만 선언되지 않아서 메모리에 저장되지 않는 함수입니다.

기본 형태는 
```javascript
let foo = function(){
console.log('foo')
}
foo()
```

와 같은 형태이고 근래에는 arrow function으로 많이 사용합니다. 
함수형 프로그래밍에서 함수를 인자로 전달할 때 많이 사용되며 
javascript 에서 주로 본 곳은 addEventListener 의 2번째 인자로 사용하는 것이나 setTimeout에서 사용하는 case 입니다.

**Comment by javascriptbiggosoo** [comment]
자바스크립트에서는 함수도 값으로 취급되기 때문에 익명 함수를 변수에 할당하거나, 콜백으로 전달하거나, 즉시 실행할 수 있습니다.

**Comment by WooGie911** [comment]
익명함수(Anonymous function)란 재사용 하지 않는, 한번만 사용할 함수를 위한 개념으로, 따로 함수의 이름을 갖지 않는 함수입니다.

따라서 재사용 될 필요가 없는 함수들의 경우 익명함수를 사용합니다. 

즉 IIFE(즉시 실행 함수)나 콜백함수, 혹은 클로저 생성을 위한 목적으로 사용하며,

익명함수가 IIFE로 사용 될 때 지역 범위 내에서 일부 코드를 캡슐화하므로 선언된 변수가 전역 범위로 누출되지 않습니다.

그리고 익명 함수는 호이스팅이 작동하지 않고, 일반 함수는 작동한다는 특징을 가지고 있습니다.

또한 보통 화살표 함수를 많이 사용하는데 기본적으로 화살표함수는 익명 함수로만 사용할 수 있기 때문에, 함수를 호출하기 위해서는 표현식으로 써야합니다.
